feat(scheduled builds): Add UI for scheduled builds proposal

JayCeeJr opened the pull request:
#### Empty
<img width="1051" alt="Screenshot 2023-05-25 at 12 52 19" src="https://github.com/go-vela/ui/assets/1265665/aa59b5be-99c8-421f-915e-9c25bf7d8d36">

#### Add Form
<img width="1053" alt="Screenshot 2023-05-25 at 12 52 29" src="https://github.com/go-vela/ui/assets/1265665/ecc00edf-3f9a-455c-9eea-075885462823">

#### List
<img width="1063" alt="Screenshot 2023-05-25 at 12 53 09" src="https://github.com/go-vela/ui/assets/1265665/6208c1a8-6ae8-4bc1-a244-4d45dabbc8b5">

JayCeeJr commented:
> is there a visual mock for what this is supposed to look like?

I added some screenshots to the description

plyr4 commented on src/elm/Pages/Schedules/Update.elm:
```diff
@@ -0,0 +1,272 @@
+
+
+
+-- takes scheduleModel and returns the URL to redirect to
+
+
+deleteScheduleRedirect : Model msg -> String
```
is `deleteScheduleRedirect` used? i cant seem to find the reference

plyr4 commented on src/elm/Pages/Schedules/Form.elm:
```diff
@@ -0,0 +1,231 @@
+
+schedulesDocsURL : String
+schedulesDocsURL =
+    "#TODO"
```
could we fill this value with velaDocsUrl?

JayCeeJr replied:
I put https://go-vela.github.io/docs/usage/schedules/ this follows the same convention as the others

plyr4 commented on src/elm/Help/Commands.elm:
```diff
@@ -716,6 +715,94 @@ addSecretArgs =
+            Just <| "vela get schedules " ++ repoArgs org repo
+
+        docs =
+            Just "/build/get"
```
```suggestion
            Just "/schedule/get"
```

plyr4 commented on src/elm/Api.elm:
```diff
@@ -698,3 +708,47 @@ deleteSecret : PartialModel a -> Engine -> Type -> Org -> Key -> Name -> Request
+-- SCHEDULES
+
+
+{-| getSchedules : fetches vela builds by repository
```
```suggestion
{-| getSchedules : fetches vela schedules by repository
```

plyr4 commented on src/elm/Pages/Schedules/Update.elm:
```diff
@@ -0,0 +1,272 @@
+            schedule
+
+
+{-| onChangeAllowCommand : updates allow\_command field on schedule update
```
```suggestion
{-| onChangeAllowCommand : updates allow_command field on schedule update
```

JayCeeJr commented on src/elm/Pages.elm:
```diff
@@ -37,6 +37,8 @@ type Page
     | BuildServices Org Repo BuildNumber FocusFragment
     | BuildPipeline Org Repo BuildNumber (Maybe ExpandTemplatesQuery) (Maybe Fragment)
+    | Schedule Org Repo ScheduleName
+    | Schedules Org Repo (Maybe Pagination.Page) (Maybe Pagination.PerPage)
```
I could go down that route, I tried this since it seemed like a lot less code, and effectively has the same result.

plyr4 replied:
true, i'd lean towards having more code if it follows the pattern we have closer

wass3rw3rk reviewed (changes requested):
feedback inline; looks to be working fine. thanks for the work!

a couple of items:

### docker compose addition
we should add `VELA_SCHEDULE_ALLOWLIST: '*'` to the docker compose for the server here, just as we did in [go-vela/server](https://github.com/go-vela/server/blob/a5fc7c68a372938e5c6ed3be182a7052ea2a9f24/docker-compose.yml#L46)

### design feedback
tl;dr: no edits requested at this time, although it would be neat if some of the following could make its way in prior to the release.

i'm not sure if there are additional plans for the add/edit view for a schedule, but it's a bit bare and geared towards advanced users at the moment. we can probably agree that cron is not the most approachable format for scheduling to start with, so i was really hoping that we can be a bit more helpful here. you will see that a few online editors stuff the raw cron expression under an "Advanced" section while the initial exposure is something more approachable (or there are helpful annotations or links to dedicated cron tools). here is some quick reference material:
- https://crontab.guru/
- https://tool.crontap.com/
- https://crontab.cronhub.io/
- https://github.com/Marak/cron-editor
- https://esarmientoe.github.io/cron-editor/
- https://karoletrych.github.io/vue-cron-editor/
- ![image](https://github.com/go-vela/ui/assets/49894298/48a2420d-096f-4f49-85b0-a7a5a99a47b9)
  (ignore the purple banner in the middle)
- ![image](https://github.com/go-vela/ui/assets/49894298/35f937d6-a809-4787-83f5-0d6830f7c27d)

not saying any of these are perfect but there are some themes/ideas. i think it boils down to these points for the current implementation:
- what is "Entry"? What am I looking at? *
- either a link to an editor or more inline explanation what each field represents, etc *
- quick views to edit common daily, weekly, hourly, etc schedules in more human way
- a cron expression to word translator would be nice
- something that tells you when the current expression would run next (taking into account server time zone) *
- calendar view would be neat (but value-add is probably diminishing there?)

maybe we can add some of the items at minimum and let user feedback guide the rest. i put a `*` next to the items that i view as lowest effort (honoring server timezone could be a little tricky?).

thoughts?

JayCeeJr commented:
> 

I changed entry to "Cron Expression" and added a help link to https://crontab.guru/
Maybe it would be better to add the cron expression explanation to the vela docs, and put links to the other cron sites there?

wass3rw3rk commented:
@JayCeeJr thanks for the edits.. agree - it should probably have a place in the docs as well.

wass3rw3rk reviewed (changes requested):
missing cypress tests still

plyr4 commented on src/elm/Main.elm:
```diff
@@ -437,6 +338,7 @@ type Msg
     | HooksResponse (Result (Http.Detailed.Error String) ( Http.Metadata, Hooks ))
     | RedeliverHookResponse Org Repo HookNumber (Result (Http.Detailed.Error String) ( Http.Metadata, String ))
     | BuildResponse Org Repo (Result (Http.Detailed.Error String) ( Http.Metadata, Build ))
+    | SchedulesResponse Org Repo (Result (Http.Detailed.Error String) ( Http.Metadata, Schedules ))
```
can we move this to below the `-- Schedules` section?

plyr4 commented on src/elm/Main.elm:
```diff
@@ -2015,6 +1932,97 @@ update msg model =
         NoOp ->
             ( model, Cmd.none )
 
+        SchedulesResponse org repo result ->
```
could we move this section of inbound HTTP responses into the `-- Inbound HTTP responses` section? 

they are currently placed in the `-- Other` section

plyr4 commented on src/elm/Main.elm:
```diff
@@ -3463,6 +3517,19 @@ loadRepoSecretsPage model maybePage maybePerPage engine org repo =
     loadRepoSubPage model org repo <| Pages.RepoSecrets engine org repo maybePage maybePerPage
 
 
+{-| loadRepoSchedulesPage : takes model org and repo and loads the page for managing repo secrets
```
```suggestion
{-| loadRepoSchedulesPage : takes model org and repo and loads the page for managing repo schedules
```

plyr4 commented on src/elm/Main.elm:
```diff
@@ -3618,6 +3685,55 @@ loadAddRepoSecretPage model engine org repo =
     )
 
 
+{-| loadAddRepoSecretPage : takes model engine org and repo and loads the page for adding secrets
```
```suggestion
{-| loadAddSchedulePage : takes model engine org and repo and loads the page for adding schedules
```

plyr4 commented on src/elm/Main.elm:
```diff
@@ -3618,6 +3685,55 @@ loadAddRepoSecretPage model engine org repo =
+    )
+
+
+{-| loadAddRepoSecretPage : takes model engine org and repo and loads the page for adding secrets
```
```suggestion
{-| loadEditSchedulePage : takes model engine org and repo and loads the page for adding schedules
```

plyr4 commented on src/elm/Main.elm:
```diff
@@ -2015,6 +1932,97 @@ update msg model =
+                    ( model, addError error )
+
+        AddScheduleResponse response ->
+            --TODO Handle Add
```
could this be removed?

plyr4 commented:
thank you for the tweaks, looks good
